server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/ObjectService.scala

```
@@ -30,21 +43,80 @@ class ObjectService(googleStorageService: GoogleStorageService[IO]) extends Http
+                            fileBody <- googleStorageService.getObject(bucketAndObject.bucketName, bucketAndObject.blobName, Some(traceId)).compile.to[Array]
+                            calculatedCrc32 = Crc32c.calculateCrc32c(fileBody)
+                            syncStatus = if (fileBody.length == 0) SyncStatus.DeletedRemote
+                              else if (calculatedCrc32 == crc32) SyncStatus.Live
+                              else SyncStatus.Desynchronized
+                            lastEditedBy <- IO.fromEither(userDefinedMetadata.get("lastEditedBy").map(WorkbenchEmail).toRight(NotFoundException("lastEditedBy key missing from object metadata ")))
+                            lastEditedTimeString <- IO.fromEither(userDefinedMetadata.get("lastEditedTime").toRight(NotFoundException("lastEditedTime key missing from object metadata ")))
+                            lastEditedTime <- IO(Instant.ofEpochMilli(lastEditedBy.value.toLong))
+                          } yield MetadataResponse(isLinked, syncStatus, lastEditedBy, lastEditedTime, sl.cloudStorageDirectory, true, sl) //TODO: fix isExecutionMode
+                      }
+          } yield result
+      }
+    } yield res
 
-  def safeDelocalize(req: SafeDelocalize): IO[Unit] = ???
+  def safeDelocalize(req: SafeDelocalize): IO[Unit] = {
+    for {
+      traceId <- IO(TraceId(randomUUID()))
+      storageLinks <- readJsonFileToA[IO, List[StorageLink]](pathToStorageLinks).compile.lastOrError
+      storageLink = storageLinks.find(_.localBaseDirectory == req.localObjectPath)
+      metadata <- checkMetadata(GetMetadataRequest(req.localObjectPath))
+      _ <- if(metadata.syncStatus == SyncStatus.Live) {
+        io.file.readAll[IO](Paths.get(req.localObjectPath.asString), blockingEc, 4096).compile.to[Array].flatMap {
+          body =>
+            googleStorageService.storeObject(GcsBucketName(""), GcsBlobName(""), body, "text/plain", Some(traceId))
+        }
+      } else {
```

nit: this should probably throw an error

---

this is not nit :)

I was coding based on assumption here and should've left a TODO fo confirming if this is what we want